daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/Parser.scala

```diff
@@ -90,6 +85,53 @@ sealed trait Parser extends Processor {
     )
   }
 
+  def PENotEnoughBits(
+    pstate: PState,
+    sfl: SchemaFileLocation,
+    dataLoc: DataLocation,
+    neededBits: Long,
+    source: InputSourceDataInputStream,
+  ): Unit = {
+    val startPos = dataLoc.bitPos1b - 1
+    val remainingLimitedBits = {
+      if (source.bitLimit0b.isEmpty) MaybeULong.Nope
+      else {
+        val lim = source.bitLimit0b.get
+        Assert.invariant(lim >= 0)
+        val nBits = lim - startPos
+        MaybeULong(nBits)
+      }
+    }
+
+    val remainingBits = {
+      if (source.hasReachedEndOfData) {
+        val bitsAvailable = {
+          val fragmentBitsRead = startPos % 8
+          source.knownBytesAvailable * 8 - fragmentBitsRead
+        }
+        if (remainingLimitedBits.isEmpty || bitsAvailable < remainingLimitedBits.get)
+          MaybeULong(bitsAvailable)
+        else
+          remainingLimitedBits
+      } else remainingLimitedBits
+    }
```

Comment:
I wonder if `remainingBits` should be a `ULong` instead of a `MaybeULong` now? If we ever call `PENotEnoughBits`, then must have hit the bit limit and `remainingLimitedBits` must be defined, or we hit the end of data, and bitsAvailable must be defined. Either way, we always have some number that represents the number of available bits. So maybe this last part just wants to be:


```scala
     {
        ...
        if (remainingLimitedBits.isEmpty || bitsAvailable < remainingLimitedBits.get)
          bitsAvailable
        else
          remainingLimitedBits.getULong
      } else remainingLimitedBits.getULong
```

And if you do the startPos adjustment at the end, then that becomes a bit cleaner since remainingBits is just a `ULong` instead of a `MaybeULong` and you don't have to check if it's defined.

Reply:
Updated